### PR TITLE
Annotate pyzx-related modules

### PR DESCRIPTION
**Pull request by EarlMilktea**
**Description of the change:**

- Check `pyzx.py` and `test_pyzx.py` by `mypy` / `pyright`
- Fix type warnings

### CODE REVIEW

**Review by wlcsm — APPROVED** [review]
lgtm

**Comment by shinich1** [comment]
@EarlMilktea Thanks,
- What's the specific reason to use `fractions`?
- `import pyzx as zx` is like `import numpy as np`, and makes it easy to. know which class/function is coming from where (graphix/pyzx). Are there strong reasons to switch to current way of writing (beyond just to suppress `mypy` warning)?

**Comment by EarlMilktea** [comment]
> What's the specific reason to use `fractions`?

This is based on the annotation in `pyzx`.
I'm not sure why this is necessary, but the annotation does reject `float`.

**Comment by EarlMilktea** [comment]
> Are there strong reasons to switch to current way of writing?

Of course yes.

Sadly the above-mentioned approach (top-level module abbreviation) works correctly in very limited situations:

1. There are no other abbreviated imports

For example, `import numpy as np` followed by `import np.random` is bad because `import` reordering ruins the module.

2. `__all__` is correctly used

Without `__all__`, we cannot determine whether an item is imported as a part of the module exports or simply for implementing the module: this is why `mypy` warns.
When `__all__` is missing, the only legal way is to directly import from the original source position.

3. Submodules are eagerly exported in `__init__.py`

For example, we can use `np.random.rand` only with `import numpy as np` because `random` is explicitly imported and listed in the `__all__` attribute of `numpy` module.
However, for most cases, we need to use something like `import numpy.random as XXX` rather than `import numpy as np` as `random` is not automatically imported.

To sum up, as `pyzx` lacks special considerations required to use `import pyzx as zx`, we have to use more verbose import styles.

**Review by thierry-martinez — CHANGES_REQUESTED** [review]
I get an error with `mypy` on `VertexType.Type`, and I noticed a code with a bad complexity in `from_pyzx_graph`.

File: graphix/pyzx.py [hunk 1]
```diff
@@ -43,31 +46,31 @@ def to_pyzx_graph(og: OpenGraph) -> BaseGraph:
             "`to_pyzx_graph` is guaranteed to work only with pyzx==0.8.0 due to possible breaking changes in `pyzx`.",
             stacklevel=1,
         )
-    g = zx.Graph()
+    g = Graph()
 
     # Add vertices into the graph and set their type
-    def add_vertices(n: int, ty: zx.VertexType) -> list[zx.VertexType]:
+    def add_vertices(n: int, ty: VertexType.Type) -> list[VertexType]:
```
[the remark this answers is not in the document]
Comment (EarlMilktea) [reply graphix/pyzx.py:52]: You might forget to install `pyzx` (optional)?
Me and CI are not encountering this error.

File: tests/test_pyzx.py [hunk 2]
```diff
@@ -1,26 +1,32 @@
...
+    try:
+        import pyzx  # noqa: F401
+
+        return False
+    except ModuleNotFoundError:
+        return True
 
 
-@pytest.mark.skipif(sys.modules.get("pyzx") is None, reason="pyzx not installed")
+@pytest.mark.skipif(_pyzx_notfound(), reason="pyzx not installed")
```
[the remark this answers is not in the document]
Comment (EarlMilktea) [reply tests/test_pyzx.py:24]: Should I use `find_spec` here?

File: graphix/pyzx.py [hunk 3]
```diff
@@ -84,23 +87,30 @@ def add_vertices(n: int, ty: zx.VertexType) -> list[zx.VertexType]:
...
-        g.add_edge((map_to_pyzx[og_index], pyzx_index), zx.EdgeType.HADAMARD)
-        g.set_phase(pyzx_index, og.measurements[og_index].angle)
+        g.add_edge((map_to_pyzx[og_index], pyzx_index), EdgeType.HADAMARD)
+        g.set_phase(pyzx_index, Fraction(og.measurements[og_index].angle))
 
     return g
 
 
-def from_pyzx_graph(g: BaseGraph) -> OpenGraph:
+def _checked_float(x: FractionLike) -> float:
```
[the remark this answers is not in the document]
Comment (EarlMilktea) [reply graphix/pyzx.py:106]: > `_check_float`

I personally prefer `_checked_float` as this function also performs cast to `float` in addition to typecheck.

> `_to_float`

I feel this name does not reflect the difference between `_checked_float` and plain `float`.

File: graphix/pyzx.py [hunk 4]
```diff
@@ -120,26 +130,27 @@ def from_pyzx_graph(g: BaseGraph) -> OpenGraph:
...
     # We need to do this since the full reduce simplification can
     # leave either hadamard or plain wires on the inputs and outputs
     for inp in g.inputs():
         first_nbr = next(iter(g.neighbors(inp)))
         et = g.edge_type((first_nbr, inp))
 
-        if et == zx.EdgeType.SIMPLE:
+        if et == EdgeType.SIMPLE:
             g_nx.remove_node(inp)
             inputs = [i if i != inp else first_nbr for i in inputs]
```
[the remark this answers is not in the document]
Comment (EarlMilktea) [reply graphix/pyzx.py:147]: I'm not sure this refactoring is valid or not because I'm not experienced with `pyzx`.

@wlcsm Could you resolve this comment in another PR?

File: graphix/pyzx.py [hunk 5]
```diff
@@ -120,26 +130,27 @@ def from_pyzx_graph(g: BaseGraph) -> OpenGraph:
...
 
     measurements = {}
-    inputs = g.inputs()
-    outputs = g.outputs()
+    inputs = list(g.inputs())
+    outputs = list(g.outputs())
 
-    g_nx = nx.Graph(g.edges())
+    edges = g.edges()
+    g_nx = nx.Graph(edges)
```
[the remark this answers is not in the document]
Comment (EarlMilktea) [reply graphix/pyzx.py]: Unnecessary at all!
Fixed in b080878105956652654eb0c865522da2d4e2ba22 .

File: graphix/pyzx.py [hunk 6]
```diff
@@ -84,23 +87,30 @@ def add_vertices(n: int, ty: zx.VertexType) -> list[zx.VertexType]:
...
-        g.set_phase(pyzx_index, og.measurements[og_index].angle)
+        g.add_edge((map_to_pyzx[og_index], pyzx_index), EdgeType.HADAMARD)
+        g.set_phase(pyzx_index, Fraction(og.measurements[og_index].angle))
 
     return g
 
 
-def from_pyzx_graph(g: BaseGraph) -> OpenGraph:
+def _checked_float(x: FractionLike) -> float:
+    if not isinstance(x, (Fraction, int)):
```
[the remark this answers is not in the document]
Comment (EarlMilktea) [reply graphix/pyzx.py]: Merged in ebdea88cfc8b426d3f6ccf2829c30d957845a7b2 .

Let me use `FractionLike` for `x` as this can be `Poly` instance that cannot be converted to `float`.

File: graphix/pyzx.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (thierry-martinez) [reply graphix/pyzx.py:52]: Oops, sorry, I don't know why I had this error (`pyzx` seemed to be installed), but the error disappeared when I restarted from a fresh environment.

File: graphix/pyzx.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (thierry-martinez) [reply graphix/pyzx.py:106]: Could `_ensure_float` work? (The idea is to use an infinitive rather than a past participle.)

File: tests/test_pyzx.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (thierry-martinez) [reply tests/test_pyzx.py:24]: If using `importlib.util.find_spec('pyzx')` is correct, then it definitely looks better than defining our own `_pyzx_notfound`. However, I'm surprised we don't already use `find_spec` in other places, where we use `sys.modules.get` instead, even though it seems incorrect.

File: graphix/pyzx.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (EarlMilktea) [reply graphix/pyzx.py:106]: I still prefer `_checked_float`...

NOTE: This naming is inspired by Rust STL (see https://doc.rust-lang.org/std/primitive.u32.html#method.checked_add for example).

File: graphix/pyzx.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (EarlMilktea) [reply graphix/pyzx.py:52]: I sometimes get into similar trouble...

File: tests/test_pyzx.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (EarlMilktea) [reply tests/test_pyzx.py:20]: By the way I found an interesting API in `pytest` (though it is not suitable for typed python)

File: graphix/pyzx.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (thierry-martinez) [reply graphix/pyzx.py:106]: In the case of `checked_add`, `add` is the infinitive verb, but I think it would be better to stop bike-shedding on this and keep `_checked_float`!

**Review by thierry-martinez — APPROVED** [review]
LGTM! Thanks.

**Comment by EarlMilktea** [comment]
@shinich1 

Can you do a final check on this PR?

**Review by shinich1 — APPROVED** [review]
